AmplifyPlugins/API/AWSAPICategoryPluginTests/Operation/AWSGraphQLSubscriptionOperationCancelTests.swift

```diff
@@ -44,12 +121,101 @@ class AWSGraphQLSubscriptionOperationCancelTests: AWSAPICategoryPluginTestBase {
+
+        let receivedCompletion = expectation(description: "Received completion")
+        receivedCompletion.isInverted = true
+        let receivedFailure = expectation(description: "Received failure")
+        let receivedValue = expectation(description: "Received value for connecting")
+        receivedValue.isInverted = true
+
+        let valueListener: GraphQLSubscriptionOperation<JSONValue>.InProcessListener = { _ in
+            receivedValue.fulfill()
+        }
+
+        let completionListener: GraphQLSubscriptionOperation<JSONValue>.ResultListener = { result in
+            switch result {
+            case .failure:
+                receivedFailure.fulfill()
+                failureSemaphore.signal()
+            case .success:
+                receivedCompletion.fulfill()
+            }
+        }
 
-        waitForExpectations(timeout: 0.05)
+        let operation = apiPlugin.subscribe(
+            request: request,
+            valueListener: valueListener,
+            completionListener: completionListener
+        )
+        failureSemaphore.wait()
+
+        XCTAssert(operation.isFinished)
+        waitForExpectations(timeout: 0.3)
     }
 
+    func testCallingCancelWhileCreatingConnectionShouldCallCompletionListener() {
+        let creatingConnectionSemaphore = DispatchSemaphore(value: 0)
+        let mockSubscriptionConnectionFactory = MockSubscriptionConnectionFactory(onGetOrCreateConnection: { _, _, _ in
+            creatingConnectionSemaphore.signal()
+            sleep(5)
+            throw APIError.invalidConfiguration("something went wrong", "", nil)
+        })
+
+        setUp(subscriptionConnectionFactory: mockSubscriptionConnectionFactory)
+
+        let request = GraphQLRequest(apiName: apiName,
+                                     document: testDocument,
+                                     variables: nil,
+                                     responseType: JSONValue.self)
+
+        let receivedCompletion = expectation(description: "Received completion")
+        let receivedFailure = expectation(description: "Received failure")
+        receivedFailure.isInverted = true
+        let receivedValue = expectation(description: "Received value for connecting")
+        receivedValue.isInverted = true
+
+        let valueListener: GraphQLSubscriptionOperation<JSONValue>.InProcessListener = { _ in
+            receivedValue.fulfill()
+        }
+
+        let completionListener: GraphQLSubscriptionOperation<JSONValue>.ResultListener = { result in
+            switch result {
+            case .failure:
+                receivedFailure.fulfill()
+            case .success:
+                receivedCompletion.fulfill()
+            }
+        }
+
+        let operation = apiPlugin.subscribe(
+            request: request,
+            valueListener: valueListener,
+            completionListener: completionListener
+        )
+        creatingConnectionSemaphore.wait()
```

Comment:
my previous comments suggested to replace the semaphore with expectations since they were fulfilling expectations right before signaling the semaphore, but for this one seems like this makes sense that it's a semaphore

Reply:
I think this one can use an expectation as well